File: skills/ablate/scripts/usage_counts.py
#!/usr/bin/env python3
"""Per-element hook fire counting and last-used dating for the ablate skill.

Reads every session transcript under the transcripts root the caller passes (the running
side's `projects/` directory) and counts, per harness element, how many PreToolUse/PostToolUse
hook fires named it and the most recent date one did.

The measurement window and the rare-by-design set live here as script constants, never as
prose in SKILL.md (docs/wiki/deterministic-script-judgment.md).
"""

from __future__ import annotations

import json
import re
import sys
from collections.abc import Iterator
from datetime import date, datetime
from pathlib import Path, PurePosixPath
from typing import TypedDict

from arms import UNMEASURED
from verdict import DELETE_CANDIDATE, NEEDS_HUMAN_JUDGMENT

# The plan's contract also names "hookSpecificOutput" records. No attachment sampled in this
# session carried that key, so reading it is deferred rather than guessed at.
FIRE_EVENTS = frozenset({"PreToolUse", "PostToolUse"})

# A `command` carries the lead-in the harness invoked it through: a path running through the
# .claude directory, or one starting from an unexpanded plugin variable. A harness element is
# named repo-root-relative, so without dropping that lead-in neither RARE_BY_DESIGN nor
# harness_elements' population matches a single key.
_CLAUDE_DIR_MARKER = "/.claude/"
_VARIABLE_PREFIX_RE = re.compile(r"^\$\{[A-Z_]+\}/")

# The suffixes that count as an element. Some fires carry a label instead of a path
# (measured in this session's transcripts: "formatter", "gates changed", "guardrails..."),
# and a label names no harness element, so it stays out of the tally.
ELEMENT_SUFFIXES = frozenset({".py", ".sh", ".js", ".ts"})

# Safety nets exercised only on an uncommon input, where zero fires must not read as unused.
# hooks/security/rm_to_trash.ts fires only when a destructive command is attempted ("Failure
# mode: fail-closed (security enforcement)"), so most sessions never trigger it.
RARE_BY_DESIGN: frozenset[str] = frozenset({"hooks/security/rm_to_trash.ts"})

# How many days back from `now` a most-recent fire still counts as observed. Past this
# window an element reports as unmeasured, rather than keeping a stale last-used date alive.
MEASUREMENT_WINDOW_DAYS = 90


class ElementUsage(TypedDict):
    fires: int
    # ISO date (YYYY-MM-DD) of the most recent fire, or None when the element never fired.
    last_used: str | None


class UsageResult(TypedDict):
    elements: dict[str, ElementUsage]
    transcript_count: int
    date_range: dict[str, str | None]


def element_path(command: str) -> str | None:
    """The repo-root-relative path of the element `command` fired, or None when it names no
    element. An absolute path, or one still leading with an unexpanded variable, has no
    repo-root-relative form, so it too returns None."""
    text = command.strip()
    cut = text.find(_CLAUDE_DIR_MARKER)
    text = text[cut + len(_CLAUDE_DIR_MARKER) :] if cut != -1 else _VARIABLE_PREFIX_RE.sub("", text)
    if not text or text[0] in "~$/":
        return None
    if PurePosixPath(text).suffix not in ELEMENT_SUFFIXES:
        return None
    return text


def _parse_date(timestamp: str) -> date | None:
    """The calendar date a transcript timestamp ("2026-08-01T00:00:00.000Z") falls on, or
    None when the value does not start with an ISO date (one malformed record must not
    stop the read, matching report.py's per-line tolerance)."""
    try:
        return datetime.strptime(timestamp[:10], "%Y-%m-%d").date()
    except ValueError:
        return None


def _iter_fires(path: Path) -> Iterator[tuple[str, date]]:
    """Yields (element_path, fire_date) once per PreToolUse/PostToolUse fire record in one
    transcript file. A malformed or incomplete record contributes nothing rather than
    raising: another process writes the transcript while this reads it, so a partial last
    line is expected."""
    with path.open(encoding="utf-8") as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            try:
                record = json.loads(line)
            except json.JSONDecodeError:
                continue
            if not isinstance(record, dict):
                continue
            attachment = record.get("attachment")
            if not isinstance(attachment, dict):
                continue
            if attachment.get("hookEvent") not in FIRE_EVENTS:
                continue
            command = attachment.get("command")
            timestamp = record.get("timestamp")
            if not isinstance(command, str) or not isinstance(timestamp, str):
                continue
            element = element_path(command)
            if element is None:
                continue
            fire_date = _parse_date(timestamp)
            if fire_date is None:
                continue
            yield element, fire_date


def count_usage(root: Path) -> UsageResult:
    """Scans every `*.jsonl` transcript under `root` and tallies fires per element. An
    element's key is the repo-root-relative path element_path returns."""
    transcripts = sorted(root.glob("**/*.jsonl"))
    elements: dict[str, ElementUsage] = {}
    fire_dates: list[date] = []

    for transcript in transcripts:
        try:
            fires = list(_iter_fires(transcript))
        except OSError:
            # One unreadable transcript must not stop the count over the rest.
            continue
        for element, fire_date in fires:
            entry = elements.setdefault(element, {"fires": 0, "last_used": None})
            entry["fires"] += 1
            if entry["last_used"] is None or fire_date.isoformat() > entry["last_used"]:
                entry["last_used"] = fire_date.isoformat()
            fire_dates.append(fire_date)

    if fire_dates:
        date_range: dict[str, str | None] = {
            "start": min(fire_dates).isoformat(),
            "end": max(fire_dates).isoformat(),
        }
    else:
        date_range = {"start": None, "end": None}

    return {
        "elements": elements,
        "transcript_count": len(transcripts),
        "date_range": date_range,
    }


# Read top to bottom; take the first row that matches. The order is load-bearing twice over:
# RARE_BY_DESIGN sits above the zero-fires row so a rare element never reaches
# DELETE_CANDIDATE, and the last_used check sits under `fires > 0` because zero fires always
# pairs with last_used=None, which a row above would swallow into UNMEASURED and leave
# DELETE_CANDIDATE unreachable.
#
# | Condition | Verdict |
# | --- | --- |
# | path is in RARE_BY_DESIGN | NEEDS_HUMAN_JUDGMENT |
# | fires > 0 and last_used is None (inconsistent input) | UNMEASURED |
# | fires > 0 and last_used falls outside MEASUREMENT_WINDOW_DAYS | UNMEASURED |
# | fires > 0 and last_used falls inside MEASUREMENT_WINDOW_DAYS | NEEDS_HUMAN_JUDGMENT |
# | fires == 0 | DELETE_CANDIDATE |
def classify(path: str, *, fires: int, last_used: str | None, now: date) -> str:
    """Assigns one element's usage observation a verdict, per the table above. Reads
    RARE_BY_DESIGN and MEASUREMENT_WINDOW_DAYS from the module namespace rather than as
    captured defaults, so patching either at run time changes the verdict returned."""
    if path in RARE_BY_DESIGN:
        return NEEDS_HUMAN_JUDGMENT
    if fires > 0:
        if last_used is None:
            return UNMEASURED
        if (now - date.fromisoformat(last_used)).days > MEASUREMENT_WINDOW_DAYS:
            return UNMEASURED
        return NEEDS_HUMAN_JUDGMENT
    return DELETE_CANDIDATE


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: usage_counts.py <transcripts-root>", file=sys.stderr)
        return 2
    result = count_usage(Path(argv[1]))
    print(json.dumps(result, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: skills/ablate/scripts/usage_counts.ts
#!/usr/bin/env node
/// <reference types="node" />
// Usage: usage_counts.ts <transcripts-root>
//
// TypeScript port of skills/ablate/scripts/usage_counts.py: FIRE_EVENTS, ELEMENT_SUFFIXES,
// RARE_BY_DESIGN, MEASUREMENT_WINDOW_DAYS, element_path, _parse_date, _iter_fires, count_usage,
// classify, main. Constant and function names stay exactly as usage_counts.py declares them,
// the same no-camelCase convention arms.ts, verdict.ts, dr_gate.ts and enforcer_map.ts hold in
// this same directory.
//
// usage_counts.py imports UNMEASURED from arms.py and DELETE_CANDIDATE/NEEDS_HUMAN_JUDGMENT
// from verdict.py; this port carries the same import, the same reuse verdict.ts already made
// for UNMEASURED.
//
// Deviation from usage_counts.py: classify's docstring there (usage_counts.py:169) says it
// "reads RARE_BY_DESIGN and MEASUREMENT_WINDOW_DAYS from the module namespace rather than as
// captured defaults, so patching either at run time changes the verdict returned" --
// unittest.mock.patch.object rebinding a module-level name. An ESM import binding cannot be
// rebound the same way from a test file (map_all/target_files in enforcer_map.ts hit the same
// wall), so this port carries MEASUREMENT_WINDOW_DAYS as classify's own `window_days`
// parameter (defaulting to the module constant) instead: a caller drives both sides of the
// boundary by passing a different value, not by patching a binding. RARE_BY_DESIGN stays a
// module-level export, read directly, unchanged.
import { globSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { isMainModule } from "../../../workflows/_lib/entry-point.ts";
import { pythonJsonStringify } from "../../_lib/python_json.ts";
import { UNMEASURED } from "./arms.ts";
import { DELETE_CANDIDATE, NEEDS_HUMAN_JUDGMENT } from "./verdict.ts";

// The plan's contract also names "hookSpecificOutput" records. No attachment sampled in this
// session carried that key, so reading it is deferred rather than guessed at.
export const FIRE_EVENTS: ReadonlySet<string> = new Set(["PreToolUse", "PostToolUse"]);

// A `command` carries the lead-in the harness invoked it through: a path running through the
// .claude directory, or one starting from an unexpanded plugin variable. A harness element is
// named repo-root-relative, so without dropping that lead-in neither RARE_BY_DESIGN nor
// harness_elements' population matches a single key.
const _CLAUDE_DIR_MARKER = "/.claude/";
const _VARIABLE_PREFIX_RE = /^\$\{[A-Z_]+\}\//;

// The suffixes that count as an element. Some fires carry a label instead of a path, and a
// label names no harness element, so it stays out of the tally.
export const ELEMENT_SUFFIXES: ReadonlySet<string> = new Set([".py", ".sh", ".js", ".ts"]);

// Safety nets exercised only on an uncommon input, where zero fires must not read as unused.
export const RARE_BY_DESIGN: ReadonlySet<string> = new Set(["hooks/security/rm_to_trash.ts"]);

// How many days back from `now` a most-recent fire still counts as observed.
export const MEASUREMENT_WINDOW_DAYS = 90;

export interface ElementUsage {
  fires: number;
  // ISO date (YYYY-MM-DD) of the most recent fire, or null when the element never fired.
  last_used: string | null;
}

export interface UsageResult {
  elements: Record<string, ElementUsage>;
  transcript_count: number;
  date_range: { start: string | null; end: string | null };
}

/** The last "/"-separated segment's suffix, the way PurePosixPath(text).suffix reads it: the
 * trailing ".ext" when the dot sits strictly inside the segment, "" otherwise (a leading dot,
 * as in ".hidden", carries no suffix either). */
function _posix_suffix(text: string): string {
  const name = text.slice(text.lastIndexOf("/") + 1);
  const dot = name.lastIndexOf(".");
  return dot > 0 && dot < name.length - 1 ? name.slice(dot) : "";
}

/** The repo-root-relative path of the element `command` fired, or null when it names no
 * element. An absolute path, or one still leading with an unexpanded variable, has no
 * repo-root-relative form, so it too returns null. */
export function element_path(command: string): string | null {
  const trimmed = command.trim();
  const cut = trimmed.indexOf(_CLAUDE_DIR_MARKER);
  const text =
    cut !== -1
      ? trimmed.slice(cut + _CLAUDE_DIR_MARKER.length)
      : trimmed.replace(_VARIABLE_PREFIX_RE, "");
  if (!text || "~$/".includes(text[0])) {
    return null;
  }
  if (!ELEMENT_SUFFIXES.has(_posix_suffix(text))) {
    return null;
  }
  return text;
}

/** The calendar date (YYYY-MM-DD) a transcript timestamp ("2026-08-01T00:00:00.000Z") falls
 * on, or null when the value does not start with a valid ISO date (one malformed record must
 * not stop the read, matching usage_counts.py's report.py-style per-line tolerance). */
function _parse_date(timestamp: string): string | null {
  const candidate = timestamp.slice(0, 10);
  if (!/^\d{4}-\d{2}-\d{2}$/.test(candidate)) {
    return null;
  }
  const parsed = new Date(`${candidate}T00:00:00.000Z`);
  // An out-of-range calendar date (e.g. "2026-02-30") rolls forward under Date's own
  // normalization instead of throwing, so re-formatting and comparing catches it the way
  // Python's strptime rejects it by raising ValueError.
  return Number.isNaN(parsed.getTime()) || parsed.toISOString().slice(0, 10) !== candidate
    ? null
    : candidate;
}

/** Yields [element_path, fire_date] once per PreToolUse/PostToolUse fire record in one
 * transcript file. A malformed or incomplete record contributes nothing rather than raising:
 * another process writes the transcript while this reads it, so a partial last line is
 * expected. */
function _iter_fires(path: string): Iterable<[string, string]> {
  const fires: [string, string][] = [];
  for (const rawLine of readFileSync(path, "utf8").split("\n")) {
    const line = rawLine.trim();
    if (!line) continue;
    let record: unknown;
    try {
      record = JSON.parse(line);
    } catch {
      continue;
    }
    if (record === null || typeof record !== "object" || Array.isArray(record)) continue;
    const attachment = (record as Record<string, unknown>).attachment;
    if (attachment === null || typeof attachment !== "object" || Array.isArray(attachment)) {
      continue;
    }
    const attachmentFields = attachment as Record<string, unknown>;
    if (!FIRE_EVENTS.has(attachmentFields.hookEvent as string)) continue;
    const command = attachmentFields.command;
    const timestamp = (record as Record<string, unknown>).timestamp;
    if (typeof command !== "string" || typeof timestamp !== "string") continue;
    const element = element_path(command);
    if (element === null) continue;
    const fireDate = _parse_date(timestamp);
    if (fireDate === null) continue;
    fires.push([element, fireDate]);
  }
  return fires;
}

/** Scans every `*.jsonl` transcript under `root` and tallies fires per element. An element's
 * key is the repo-root-relative path element_path returns. */
export function count_usage(root: string): UsageResult {
  const transcripts = globSync("**/*.jsonl", { cwd: root }).sort();
  const elements: Record<string, ElementUsage> = {};
  let start: string | null = null;
  let end: string | null = null;

  for (const relative of transcripts) {
    let fires: Iterable<[string, string]>;
    try {
      fires = _iter_fires(join(root, relative));
    } catch {
      // One unreadable transcript must not stop the count over the rest.
      continue;
    }
    for (const [element, fireDate] of fires) {
      const entry = (elements[element] ??= { fires: 0, last_used: null });
      entry.fires += 1;
      if (entry.last_used === null || fireDate > entry.last_used) {
        entry.last_used = fireDate;
      }
      if (start === null || fireDate < start) start = fireDate;
      if (end === null || fireDate > end) end = fireDate;
    }
  }

  return {
    elements,
    transcript_count: transcripts.length,
    date_range: { start, end },
  };
}

// Read top to bottom; take the first row that matches. The order is load-bearing twice over:
// RARE_BY_DESIGN sits above the zero-fires row so a rare element never reaches
// DELETE_CANDIDATE, and the last_used check sits under `fires > 0` because zero fires always
// pairs with last_used=null, which a row above would swallow into UNMEASURED and leave
// DELETE_CANDIDATE unreachable.
//
// | Condition | Verdict |
// | --- | --- |
// | path is in RARE_BY_DESIGN | NEEDS_HUMAN_JUDGMENT |
// | fires > 0 and last_used is null (inconsistent input) | UNMEASURED |
// | fires > 0 and last_used falls outside window_days | UNMEASURED |
// | fires > 0 and last_used falls inside window_days | NEEDS_HUMAN_JUDGMENT |
// | fires == 0 | DELETE_CANDIDATE |
/** Assigns one element's usage observation a verdict, per the table above. `window_days`
 * replaces usage_counts.py's module-namespace read of MEASUREMENT_WINDOW_DAYS -- see the
 * header deviation note above. */
export function classify(
  path: string,
  fires: number,
  last_used: string | null,
  now: Date,
  window_days: number = MEASUREMENT_WINDOW_DAYS,
): string {
  if (RARE_BY_DESIGN.has(path)) {
    return NEEDS_HUMAN_JUDGMENT;
  }
  if (fires > 0) {
    if (last_used === null) {
      return UNMEASURED;
    }
    const lastUsedMs = Date.parse(`${last_used}T00:00:00.000Z`);
    const days = Math.round((now.getTime() - lastUsedMs) / (24 * 60 * 60 * 1000));
    if (days > window_days) {
      return UNMEASURED;
    }
    return NEEDS_HUMAN_JUDGMENT;
  }
  return DELETE_CANDIDATE;
}

// The wire format has to follow json.dumps' byte-for-byte spacing, not JSON.stringify's default
// -- see python_json.ts's header for why, and for the shared encoder every CLI in this family
// reuses instead of hand-building its own braces.
export function main(argv: string[]): number {
  if (argv.length !== 1) {
    process.stderr.write("usage: usage_counts.ts <transcripts-root>\n");
    return 2;
  }
  process.stdout.write(`${pythonJsonStringify(count_usage(argv[0]))}\n`);
  return 0;
}

if (isMainModule(import.meta.url)) {
  process.exit(main(process.argv.slice(2)));
}
